crates/ditto-checker/src/typechecker/mod.rs

```diff
@@ -668,6 +677,46 @@ fn infer_or_check_match(
     })
 }
 
+fn check_exhaustiveness(
+    env: &Env,
+    state: &mut State,
+    match_span: Span,
+    pattern_type: Type,
+    patterns: Vec<Pattern>,
+) -> Result<()> {
+    match coverage::is_exhaustive(&env.constructors, pattern_type, patterns) {
+        None => Ok(()),
+        Some(coverage::Error::RedundantClauses(clause_patterns)) => {
+            state
+                .warnings
+                .extend(clause_patterns.into_iter().map(|clause_pattern| {
+                    Warning::RedundantMatchPattern {
+                        span: clause_pattern.get_span(),
+                    }
+                }));
+            Ok(())
+        }
+        Some(coverage::Error::NotCovered(ideal_patterns)) => {
+            let mut missing_patterns = ideal_patterns
+                .into_iter()
+                .map(|ideal_pattern| ideal_pattern.render())
```

Comment:
I'm tempted to wrap the rendered pattern in backticks to make it clearer that it's code (rather than prose) but that might be too visually noisy?